File: libs/solvers/unittests/test_molecule.cpp
/****************************************************************-*- C++ -*-****
 * Copyright (c) 2024 NVIDIA Corporation & Affiliates.                         *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/
#include "cudaq/solvers/operators.h"
#include "cudaq/solvers/operators/molecule/molecule_package_driver.h"

#include <fstream>
#include <iostream>
#include <tuple>

#include <gtest/gtest.h>

TEST(MoleculeTester, checkSimple) {
  auto registeredNames =
      cudaq::solvers::MoleculePackageDriver::get_registered();

  EXPECT_EQ(registeredNames.size(), 1);
  EXPECT_TRUE(std::find(registeredNames.begin(), registeredNames.end(),
                        "RESTPySCFDriver") != registeredNames.end());
  {
    cudaq::solvers::molecular_geometry geometry{{"H", {0., 0., 0.}},
                                                {"H", {0., 0., .7474}}};
    auto molecule = cudaq::solvers::create_molecule(
        geometry, "sto-3g", 0, 0, {.casci = true, .verbose = true});

    molecule.hamiltonian.dump();

    EXPECT_NEAR(molecule.energies["fci_energy"], -1.137, 1e-3);
    EXPECT_NEAR(molecule.energies["hf_energy"], -1.1163255644, 1e-3);
    EXPECT_EQ(molecule.n_electrons, 2);
    EXPECT_EQ(molecule.n_orbitals, 2);

    // EXPECT_NEAR(molecule.fermionOperator.constant, 0.7080240981000804, 1e-3);
    EXPECT_EQ(2, molecule.hpq.shape().size());
    EXPECT_EQ(4, molecule.hpq.shape()[0]);
    EXPECT_EQ(4, molecule.hpq.shape()[1]);
    EXPECT_NEAR(molecule.hpq.at({0, 0}).real(), -1.2488, 1e-3);
    EXPECT_NEAR(molecule.hpq.at({1, 1}).real(), -1.2488, 1e-3);
    EXPECT_NEAR(molecule.hpq.at({2, 2}).real(), -.47967, 1e-3);
    EXPECT_NEAR(molecule.hpq.at({3, 3}).real(), -.47967, 1e-3);
    EXPECT_EQ(4, molecule.hpqrs.shape().size());
    for (int i = 0; i < 4; i++)
      EXPECT_EQ(4, molecule.hpqrs.shape()[i]);
    EXPECT_NEAR(molecule.hpqrs.at({0, 0, 0, 0}).real(), 0.3366719725032414,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({0, 0, 2, 2}).real(), 0.0908126657382825,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({0, 1, 1, 0}).real(), 0.3366719725032414,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({0, 1, 3, 2}).real(), 0.0908126657382825,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({0, 2, 0, 2}).real(), 0.09081266573828267,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({0, 2, 2, 0}).real(), 0.33121364716348484,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({0, 3, 1, 2}).real(), 0.09081266573828267,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({0, 3, 3, 0}).real(), 0.33121364716348484,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({1, 0, 0, 1}).real(), 0.3366719725032414,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({1, 0, 2, 3}).real(), 0.0908126657382825,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({1, 1, 1, 1}).real(), 0.3366719725032414,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({1, 1, 3, 3}).real(), 0.0908126657382825,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({1, 2, 0, 3}).real(), 0.09081266573828267,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({1, 2, 2, 1}).real(), 0.33121364716348484,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({1, 3, 1, 3}).real(), 0.09081266573828267,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({1, 3, 3, 1}).real(), 0.33121364716348484,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({2, 0, 0, 2}).real(), 0.3312136471634851,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({2, 0, 2, 0}).real(), 0.09081266573828246,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({2, 1, 1, 2}).real(), 0.3312136471634851,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({2, 1, 3, 0}).real(), 0.09081266573828246,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({2, 2, 0, 0}).real(), 0.09081266573828264,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({2, 2, 2, 2}).real(), 0.34814578499360427,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({2, 3, 1, 0}).real(), 0.09081266573828264,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({2, 3, 3, 2}).real(), 0.34814578499360427,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({3, 0, 0, 3}).real(), 0.3312136471634851,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({3, 0, 2, 1}).real(), 0.09081266573828246,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({3, 1, 1, 3}).real(), 0.3312136471634851,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({3, 1, 3, 1}).real(), 0.09081266573828246,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({3, 2, 0, 1}).real(), 0.09081266573828264,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({3, 2, 2, 3}).real(), 0.34814578499360427,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({3, 3, 1, 1}).real(), 0.09081266573828264,
                1e-3);
    EXPECT_NEAR(molecule.hpqrs.at({3, 3, 3, 3}).real(), 0.34814578499360427,
                1e-3);
  }
}

TEST(OperatorsTester, checkH2OActiveSpace) {
  std::string contents = R"#(3

O 0.1173 0.0 0.0
H -0.4691 0.7570 0.0
H -0.4691 -0.7570 0.0
)#";

  {
    std::ofstream out(".tmpH2O.xyz");
    out << contents;
  }

  auto geometry = cudaq::solvers::molecular_geometry::from_xyz(".tmpH2O.xyz");
  std::remove(".tmpH2O.xyz");

  auto molecule = cudaq::solvers::create_molecule(
      geometry, "631g", 0, 0,
      {.nele_cas = 6, .norb_cas = 6, .ccsd = true, .verbose = true});

  // molecule.hamiltonian.dump();
  EXPECT_EQ(molecule.n_electrons, 6);
  EXPECT_EQ(molecule.n_orbitals, 6);
}

File: libs/solvers/python/tests/test_molecule.py
# ============================================================================ #
# Copyright (c) 2024 - 2025 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

import os

import pytest, pathlib
import numpy as np
import cudaq_solvers as solvers

currentPath = pathlib.Path(__file__).parent.resolve()


def test_operators():
    geometry = [('H', (0., 0., 0.)), ('H', (0., 0., .7474))]
    molecule = solvers.create_molecule(geometry,
                                       'sto-3g',
                                       0,
                                       0,
                                       verbose=True,
                                       casci=True)
    print(molecule.hamiltonian)
    print(molecule.energies)
    assert np.isclose(-1.1163, molecule.energies['hf_energy'], atol=1e-2)
    assert np.isclose(-1.1371, molecule.energies['fci_energy'], atol=1e-4)
    from scipy.linalg import eigh
    minE = eigh(molecule.hamiltonian.to_matrix(), eigvals_only=True)[0]
    assert np.isclose(-1.1371, minE, atol=1e-4)


def test_from_xyz_filename():
    molecule = solvers.create_molecule(str(currentPath) + '/resources/LiH.xyz',
                                       'sto-3g',
                                       0,
                                       0,
                                       verbose=True)
    print(molecule.energies)
    print(molecule.n_orbitals)
    print(molecule.n_electrons)
    assert molecule.n_orbitals == 6
    assert molecule.n_electrons == 4


def test_jordan_wigner():
    geometry = [('H', (0., 0., 0.)), ('H', (0., 0., .7474))]
    molecule = solvers.create_molecule(geometry,
                                       'sto-3g',
                                       0,
                                       0,
                                       verbose=True,
                                       casci=True)

    op = solvers.jordan_wigner(molecule.hpq,
                               molecule.hpqrs,
                               core_energy=molecule.energies['nuclear_energy'],
                               tol=1e-15)
    assert molecule.hamiltonian == op
    hpq = np.array(molecule.hpq)
    hpqrs = np.array(molecule.hpqrs)
    hpqJw = solvers.jordan_wigner(hpq, molecule.energies['nuclear_energy'])
    hpqrsJw = solvers.jordan_wigner(hpqrs)
    op2 = hpqJw + hpqrsJw
    assert np.allclose(op2.to_matrix(),
                       molecule.hamiltonian.to_matrix(),
                       atol=1e-12)

    spin_ham_matrix = molecule.hamiltonian.to_matrix()
    e, c = np.linalg.eig(spin_ham_matrix)
    assert np.isclose(np.min(e), -1.13717, rtol=1e-4)

    spin_ham_matrix = op2.to_matrix()
    e, c = np.linalg.eig(spin_ham_matrix)
    assert np.isclose(np.min(e), -1.13717, rtol=1e-4)


def test_bravyi_kitaev():
    geometry = [('H', (0., 0., 0.)), ('H', (0., 0., .7474))]
    molecule = solvers.create_molecule(geometry,
                                       'sto-3g',
                                       0,
                                       0,
                                       verbose=True,
                                       casci=True)

    op = solvers.bravyi_kitaev(molecule.hpq,
                               molecule.hpqrs,
                               core_energy=molecule.energies['nuclear_energy'],
                               tol=1e-15)
    spin_ham_matrix = molecule.hamiltonian.to_matrix()
    e, c = np.linalg.eig(spin_ham_matrix)
    assert np.isclose(np.min(e), -1.13717, rtol=1e-4)

    spin_ham_matrix = op.to_matrix()
    e, c = np.linalg.eig(spin_ham_matrix)
    assert np.isclose(np.min(e), -1.13717, rtol=1e-4)


def test_active_space():

    geometry = [('N', (0.0, 0.0, 0.5600)), ('N', (0.0, 0.0, -0.5600))]
    molecule = solvers.create_molecule(geometry,
                                       'sto-3g',
                                       0,
                                       0,
                                       nele_cas=4,
                                       norb_cas=4,
                                       ccsd=True,
                                       casci=True,
                                       verbose=True)
    assert molecule.n_orbitals == 4
    assert molecule.n_electrons == 4
    assert np.isclose(molecule.energies['core_energy'], -102.139973, rtol=1e-4)
    assert np.isclose(molecule.energies['R-CCSD'], -107.5421878, rtol=1e-4)
    assert np.isclose(molecule.energies['R-CASCI'], -107.5421983, rtol=1e-4)

    print(molecule.energies)
    print(molecule.n_orbitals)
    print(molecule.n_electrons)


def test_jordan_wigner_as():
    geometry = [('N', (0.0, 0.0, 0.5600)), ('N', (0.0, 0.0, -0.5600))]
    molecule = solvers.create_molecule(geometry,
                                       'sto-3g',
                                       0,
                                       0,
                                       nele_cas=4,
                                       norb_cas=4,
                                       ccsd=True,
                                       casci=True,
                                       verbose=True)

    op = solvers.jordan_wigner(molecule.hpq, molecule.hpqrs,
                               molecule.energies['core_energy'])

    print(op)
    assert molecule.hamiltonian == op

    hpq = np.array(molecule.hpq)
    hpqrs = np.array(molecule.hpqrs)
    hpqJw = solvers.jordan_wigner(hpq,
                                  core_energy=molecule.energies['core_energy'],
                                  tolerance=1e-15)
    hpqrsJw = solvers.jordan_wigner(hpqrs)
    op2 = hpqJw + hpqrsJw

    spin_ham_matrix = molecule.hamiltonian.to_matrix()
    e, c = np.linalg.eig(spin_ham_matrix)
    print(np.min(e))
    assert np.isclose(np.min(e), -107.542198, rtol=1e-4)

    spin_ham_matrix = op2.to_matrix()
    e, c = np.linalg.eig(spin_ham_matrix)
    print(np.min(e))
    assert np.isclose(np.min(e), -107.542198, rtol=1e-4)


def test_bravyi_kitaev_as():
    geometry = [('N', (0.0, 0.0, 0.5600)), ('N', (0.0, 0.0, -0.5600))]
    molecule = solvers.create_molecule(geometry,
                                       'sto-3g',
                                       0,
                                       0,
                                       nele_cas=4,
                                       norb_cas=4,
                                       ccsd=True,
                                       casci=True,
                                       verbose=True)

    op = solvers.bravyi_kitaev(molecule.hpq,
                               molecule.hpqrs,
                               core_energy=molecule.energies['core_energy'],
                               tol=1e-15)
    spin_ham_matrix = molecule.hamiltonian.to_matrix()
    e, c = np.linalg.eig(spin_ham_matrix)
    assert np.isclose(np.min(e), -107.542198, rtol=1e-4)

    spin_ham_matrix = op.to_matrix()
    e, c = np.linalg.eig(spin_ham_matrix)
    assert np.isclose(np.min(e), -107.542198, rtol=1e-4)


def test_as_with_natorb():

    geometry = [('N', (0.0, 0.0, 0.5600)), ('N', (0.0, 0.0, -0.5600))]
    molecule = solvers.create_molecule(geometry,
                                       'sto-3g',
                                       0,
                                       0,
                                       nele_cas=4,
                                       norb_cas=4,
                                       MP2=True,
                                       ccsd=True,
                                       casci=True,
                                       natorb=True,
                                       integrals_natorb=True,
                                       verbose=True)
    assert molecule.n_orbitals == 4
    assert molecule.n_electrons == 4
    assert np.isclose(molecule.energies['R-CCSD'], -107.6059540, rtol=1e-4)
    assert np.isclose(molecule.energies['R-CASCI'], -107.6076127, rtol=1e-4)

    print(molecule.energies)
    print(molecule.n_orbitals)
    print(molecule.n_electrons)


def test_as_with_casscf():

    geometry = [('N', (0.0, 0.0, 0.5600)), ('N', (0.0, 0.0, -0.5600))]
    molecule = solvers.create_molecule(geometry,
                                       'sto-3g',
                                       0,
                                       0,
                                       nele_cas=4,
                                       norb_cas=4,
                                       MP2=True,
                                       ccsd=True,
                                       casci=True,
                                       casscf=True,
                                       natorb=True,
                                       integrals_casscf=True,
                                       verbose=True)

    assert molecule.n_orbitals == 4
    assert molecule.n_electrons == 4
    assert np.isclose(molecule.energies['R-CASSCF'], -107.607626, rtol=1e-4)

    print(molecule.energies)
    print(molecule.n_orbitals)
    print(molecule.n_electrons)


def test_H2_UR():
    geometry = [('H', (0., 0., 0.)), ('H', (0., 0., .7474))]
    molecule = solvers.create_molecule(geometry,
                                       'sto-3g',
                                       0,
                                       0,
                                       UR=True,
                                       ccsd=True,
                                       casci=True,
                                       verbose=True)

    print(molecule.energies)
    assert np.isclose(-1.1163, molecule.energies['hf_energy'], atol=1e-4)
    assert np.isclose(-1.1371, molecule.energies['fci_energy'], atol=1e-4)
    from scipy.linalg import eigh
    minE = eigh(molecule.hamiltonian.to_matrix(), eigvals_only=True)[0]
    assert np.isclose(-1.1371, minE, atol=1e-4)


def test_N2_UR_as():
    geometry = [('N', (0., 0., 0.56)), ('N', (0., 0., -0.56))]
    molecule = solvers.create_molecule(geometry,
                                       '631g',
                                       0,
                                       0,
                                       UR=True,
                                       nele_cas=4,
                                       norb_cas=4,
                                       ccsd=True,
                                       casci=True,
                                       verbose=True)

    print(molecule.energies)
    assert molecule.n_orbitals == 4
    assert molecule.n_electrons == 4
    assert np.isclose(molecule.energies['core_energy'], -103.31815, atol=1e-5)
    assert np.isclose(molecule.energies['UR-CCSD'], -108.942725, atol=1e-5)
    assert np.isclose(molecule.energies['UR-CASCI'], -108.94365, atol=1e-5)

    op = solvers.jordan_wigner(molecule.hpq, molecule.hpqrs,
                               molecule.energies['core_energy'])

    assert molecule.hamiltonian == op

    from scipy.linalg import eigh
    minE = eigh(molecule.hamiltonian.to_matrix(), eigvals_only=True)[0]
    assert np.isclose(minE, -108.9436, atol=1e-4)
